### PR TITLE
Initial structure of telemetry feature in the extension

### PR DESCRIPTION
**Pull request by Achal1607**
Telemetry feature would help us collect extension usage information to improve our extension. Usage info would only be collected if user opts in for the same.

### CODE REVIEW

**Review by sid-srini — COMMENTED** [review]
Thanks a lot @Achal1607.

File: vscode/src/telemetry/events/baseEvent.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+    public onSuccessPostEventCallback = async (): Promise<void> => {
+        LOGGER.debug(`${this.NAME} sent successfully`);
+    }
+
+    public onFailPostEventCallback = async (): Promise<void> => {
+        LOGGER.debug(`${this.NAME} send failed`);
+    }
+
+    protected addEventToCache = (): void => {
```
[the remark this answers is not in the document]
Comment (Achal1607) [reply vscode/src/telemetry/events/baseEvent.ts:58]: Yes your understanding is correct, max would be equal to number of event-types

File: vscode/src/telemetry/impl/postTelemetry.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+export interface TelemetryPostResponse {
+    success: TelemetryEventResponse[];
+    failures: TelemetryEventResponse[];
+};
+
+export class PostTelemetry {
+    public post = async (events: BaseEvent<any>[]): Promise<TelemetryPostResponse> => {
+        try {
+            if (TELEMETRY_API.baseEndpoint == null) {
```
Comment (sid-srini) [inline vscode/src/telemetry/impl/postTelemetry.ts:33]: Perhaps this is supposed to be:
```suggestion
            if (TELEMETRY_API.baseUrl == null) {
```

File: vscode/src/telemetry/telemetry.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import { LOGGER } from "../logger";
+import { BaseEvent } from "./events/baseEvent";
+import { TelemetryReporter } from "./types";
+import { TELEMETRY_API } from "./config";
+
+export namespace Telemetry {
+
+	let telemetryManager: TelemetryManager;
+	
+	export const isTelemetryFeatureAvailable = TELEMETRY_API.baseEndpoint != null && TELEMETRY_API.baseEndpoint.trim().length;
```
Comment (sid-srini) [inline vscode/src/telemetry/telemetry.ts:27]: Perhaps this is supposed to be:
```suggestion
	export const isTelemetryFeatureAvailable = TELEMETRY_API.baseUrl != null && TELEMETRY_API.baseUrl.trim().length;
```

**Review by sid-srini — CHANGES_REQUESTED** [review]

**Review by sid-srini — APPROVED** [review]
Thanks a lot @Achal1607. LGTM 👍

**Pull request merged**